File: labs/06-producer-consumer-pipeline/python/main.py
from collections import deque
from threading import Thread, Condition


def run(producer_count=1, consumer_count=1, queue_size=5, items_to_produce=10):
    if producer_count <= 0 or consumer_count <= 0 or queue_size <= 0:
        raise ValueError("producer_count, consumer_count, and queue_size must be positive")
    if items_to_produce < 0:
        raise ValueError("items_to_produce must be non-negative")

    queue = deque()
    condition = Condition()
    producer_done = False

    def producer():
        for item in range(items_to_produce):
            with condition:
                while len(queue) == queue_size:
                    condition.wait()
                queue.append(item)
                condition.notify_all()

    def consumer(consumer_id):
        while True:
            with condition:
                while not queue and not producer_done:
                    condition.wait()
                if not queue:
                    return
                item = queue.popleft()
                condition.notify_all()
            print(f"Consumer {consumer_id} consumed item: {item}")

    producers = [Thread(target=producer) for _ in range(producer_count)]
    consumers = [Thread(target=consumer, args=(i,)) for i in range(consumer_count)]
    for p in producers:
        p.start()
    for c in consumers:
        c.start()
    for p in producers:
        p.join()
    # Signal consumers to stop after all producers are done
    with condition:
        producer_done = True
        condition.notify_all()
    for c in consumers:
        c.join()

File: labs/06-producer-consumer-pipeline/go/main.go
package main

import (
	"context"
	"errors"
	"sync"
)

// runPipeline sends numItems from each producer to the consumer workers.
// Context cancellation stops remaining work; consumer callbacks must not fail.
func runPipeline(ctx context.Context, numProducers, numConsumers, bufferSize, numItems int, consume func(int)) error {
	if ctx == nil {
		return errors.New("context must not be nil")
	}
	if numProducers <= 0 || numConsumers <= 0 || bufferSize <= 0 || numItems < 0 {
		return errors.New("producer count, consumer count, and buffer size must be positive; item count must be non-negative")
	}
	if consume == nil {
		return errors.New("consumer function must not be nil")
	}

	jobs := make(chan int, bufferSize)
	var producers, consumers sync.WaitGroup

	for range numProducers {
		producers.Add(1)
		go func() {
			defer producers.Done()
			for j := range numItems {
				select {
				case <-ctx.Done():
					return
				case jobs <- j:
				}
			}
		}()
	}

	for range numConsumers {
		consumers.Add(1)
		go func() {
			defer consumers.Done()
			for {
				select {
				case <-ctx.Done():
					return
				case job, ok := <-jobs:
					if !ok {
						return
					}
					consume(job)
				}
			}
		}()
	}

	producers.Wait()
	close(jobs)
	consumers.Wait()

	return ctx.Err()
}
